### DISCUSSION TITLE
Execute server-only code inside site.ts

### DISCUSSION THREAD

**Discussion by marcopixel** [ElMassimo/iles#99] [Q&A] [answered]
Hello,

first of all - thank you for your awesome project 👍 
It's really easy to use and works perfectly for my needs as a static site generator :)

Only thing i've haven't found yet is if there's a way you can run server-only functions like `fs` or `child_process` inside the site.ts file?
Seems like Vite is trying to run it in the browser which obviously doesn't work and i haven't found anything in the docs/codebase which could help me with this issue.

Here's an example down below (using child_process to get latest GIT hash):

```ts
import { execSync } from "child_process";

const commitHash = execSync('git rev-parse --short HEAD').toString().trim();

export default {
  title: 'îles',
  description: 'Islands of interactivity with Vue in Vite.js',
  commitHash
}
```

Thanks in advance! 😄

**Comment by ElMassimo** [comment, marked as answer]
Hi @marcopixel!

You can run server-only code behind a `import.meta.env.SSR` guard. For example:

```ts
let commitHash

if (import.meta.env.SSR) {
  const { execSync } = require('child_process')
  commitHash = execSync('git rev-parse --short HEAD').toString().trim()
}
```

However, if you want a more "universal" approach, I would recommend using an env var with [`define`](https://vitejs.dev/config/#define), for example in `iles.config.ts` where you can use node directly:

```ts
export default defineConfig({
  vite: {
    define: {
      'import.meta.env.APP_GIT_SHA': JSON.stringify(commitHash),
    },
  },
})
```

---

Just in case, here's a related example, but [setting git timestamps to the frontmatter of each page](https://github.com/ElMassimo/iles/blob/main/docs/modules/lastUpdated.ts#L9).

**Reply by marcopixel** [reply]
Wow, thanks for the fast reply.

I've went with the define variant (forgot it existed 😅) and it works perfectly! 👍
